code/go/0chain.net/smartcontract/zcnsc/stakepool.go

```diff
@@ -0,0 +1,313 @@
+package zcnsc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	cstate "0chain.net/chaincore/chain/state"
+	"0chain.net/chaincore/state"
+	"0chain.net/chaincore/transaction"
+	"0chain.net/core/common"
+	"0chain.net/core/datastore"
+	"0chain.net/core/encryption"
+	"0chain.net/core/util"
+	"0chain.net/smartcontract/stakepool"
+	"0chain.net/smartcontract/stakepool/spenum"
+)
+
+// unlock response
+type unlockResponse struct {
+	// one of the fields is set in a response, the Unstake if can't unstake
+	// for now and the TokenPoolTransferResponse if it has a pool had unlocked
+	Unstake bool          `json:"unstake"` // max time to wait to unstake
+	Balance state.Balance `json:"balance"`
+}
+
+type stakePoolRequest struct {
+	PoolID       string `json:"pool_id,omitempty"`
+	AuthorizerID string `json:"authorizer_id,omitempty"`
+}
+
+func (spr *stakePoolRequest) decode(p []byte) (err error) {
+	if err = json.Unmarshal(p, spr); err != nil {
+		return
+	}
+	return // ok
+}
+
+// ----------- LockingPool pool --------------------------
+
+type StakePool struct {
+	stakepool.StakePool
+}
+
+func newStakePool() *StakePool {
+	bsp := stakepool.NewStakePool()
+	return &StakePool{
+		StakePool: *bsp,
+	}
+}
+
+// stake pool key for the storage SC and service provider ID
+func stakePoolKey(scKey, providerID string) datastore.Key {
+	return scKey + ":stakepool:" + providerID
+}
+
+func stakePoolID(scKey, providerID string) datastore.Key {
+	return encryption.Hash(stakePoolKey(scKey, providerID))
+}
+
+// Encode to []byte
+func (sp *StakePool) Encode() (b []byte) {
+	var err error
+	if b, err = json.Marshal(sp); err != nil {
+		panic(err) // must never happen
+	}
+	return
+}
+
+// Decode from []byte
+func (sp *StakePool) Decode(input []byte) error {
+	return json.Unmarshal(input, sp)
+}
+
+// save the stake pool
+func (sp *StakePool) save(sscKey, providerID string, balances cstate.StateContextI) (err error) {
+	_, err = balances.InsertTrieNode(stakePoolKey(sscKey, providerID), sp)
+	return
+}
+
+// The stake() returns total stake size including delegate pools want to unstake.
```

Comment:
The base class has a [stake()](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/stakepool/stakepool.go#L215) method, probably a better idea to make it public.